client/core/core.go

```diff
@@ -2397,20 +2629,107 @@ func (c *Core) verifyRegistrationFee(assetID uint32, dc *dexConnection, coinID [
 
 // IsInitialized checks if the app is already initialized.
 func (c *Core) IsInitialized() (bool, error) {
-	return c.db.ValueExists(keyParamsKey)
+	if c.creds() != nil {
+		return true, nil
+	}
+	// Handle the case of a Core version 0 with legacy credentials, before
+	// the login-upgrade.
+	if ver, err := c.db.CoreVersion(); err == nil && ver == 0 {
+		if _, err := c.db.LegacyKeyParams(); err == nil {
+			return true, nil
+		}
+	}
+	return false, nil
 }
 
-// InitializeClient sets the initial app-wide password for the client.
-func (c *Core) InitializeClient(pw []byte) error {
+// InitializeClient sets the initial app-wide password and app seed for the
+// client. The seed argument should be left nil unless restoring from seed.
+func (c *Core) InitializeClient(pw, restorationSeed []byte) error {
 	if initialized, err := c.IsInitialized(); err != nil {
 		return fmt.Errorf("error checking if app is already initialized: %w", err)
 	} else if initialized {
 		return fmt.Errorf("already initialized, login instead")
 	}
 
-	// Set app password.
-	_, err := c.setAppPass(pw)
-	return err
+	_, creds, err := c.generateCredentials(pw, restorationSeed)
+	if err != nil {
+		return err
+	}
+
+	err = c.db.SetPrimaryCredentials(creds, Version)
+	if err != nil {
+		return fmt.Errorf("SetPrimaryCredentials error: %w", err)
+	}
+	c.setCredentials(creds)
+
+	if len(restorationSeed) == 0 {
+		msg := "A new application seed has been created. Make a back up now in the settings view."
+		c.notify(newSecurityNote(SubjectSeedNeedsSaving, msg, db.Success))
+	}
+
+	return nil
+}
+
+// ExportSeed exports the application seed.
+func (c *Core) ExportSeed(pw []byte) ([]byte, error) {
+
+	crypter, err := c.encryptionKey(pw)
+	if err != nil {
+		return nil, fmt.Errorf("ExportSeed password error: %w", err)
+	}
+
+	creds := c.creds()
+	if creds == nil {
+		return nil, fmt.Errorf("no v2 credentials stored")
+	}
+
+	seed, err := crypter.Decrypt(creds.EncSeed)
+	if err != nil {
+		return nil, fmt.Errorf("seed decryption error: %w", err)
+	}
+
+	return seed, nil
+}
+
+// generateCredentials generates a new set of *PrimaryCredentials. The
+// credentials are not stored to the database.
+func (c *Core) generateCredentials(pw, restorationSeed []byte) (encrypt.Crypter, *db.PrimaryCredentials, error) {
+	if len(pw) == 0 {
+		return nil, nil, fmt.Errorf("empty password not allowed")
+	}
+
+	// Generate an inner key and it's Crypter.
+	innerKey := encode.RandomBytes(32)
+	innerCrypter := c.newCrypter(innerKey)
+
+	// Generate the outer key.
+	outerCrypter := c.newCrypter(pw)
+	encInnerKey, err := outerCrypter.Encrypt(innerKey)
+	if err != nil {
+		return nil, nil, fmt.Errorf("inner key encryption error: %w", err)
+	}
+
+	// Generate a seed to use as the root for all future key generation.
+	seed := restorationSeed
+	if len(seed) == 0 {
+		seed = encode.RandomBytes(64)
+	} else if len(seed) != 64 {
```

Comment:
The 512-bit/64-byte characteristic should probable be set with a constant since DB also requires this

Reply:
Constant is good, but DB never actually handles the decrypted seed.